Implement cellCharter

Jieran-S opened the pull request:
Cellcharter is implemented and tested. Should be ready to merge #7

Jieran-S committed:
Implement cellCharter

Jieran-S committed:
remove pycache

niklasmueboe commented on method/CellCharter/CellCharter.py:
```diff
@@ -0,0 +1,195 @@
+import scanpy as sc
+import pandas as pd
+import numpy as np
+from squidpy._constants._pkg_constants import Key
```
importing from private modules seems dangerous. Also I can't find this used anywhere?

Probably tis import should be removed

Jieran-S replied:
Hi yes sorry this import will be removed!

niklasmueboe commented on method/CellCharter/CellCharter.py:
```diff
@@ -0,0 +1,195 @@
+adata.uns["spatial_neighbors"] = {
+        "connectivities_key": "spatial_connectivities",
+        "distances_key": "spatial_distances",
+        "params": {"coord_type": "generic", "delaunay": True, "transform": None},
```
These params might not always be correct. Is this used in later steps of the algorithm?

Jieran-S replied:
Hi, these parameters aren't really used in any of the downstream analyses. I simply put in there as the code for calculating the spatial distance is taken from the case where `delaunay=True` and `coord_type=generic` in [`squidpy.gr.spatial_neighbors`](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_neighbors.html). 

The main point of code lines 143-159 is to replicate how the squidpy function generates the distance matrix from the adjacency matrix. And here is what each parameter is for: 
* `delaunay=True`: It is set to be `True` because in cellcharter tutorial they mentioned the delaunay parameters need to be true for downstream functions. I also adapted the code for the distance matrix assuming `delaunay=True` in the function.
* `coord_type=generic`: If `delaunay=True`, this parameter doesn't make a difference, set to default
* `transform=None`: Set to default, assuming no transformation on the adjacency matrix.

Hope it answers the question! Those parameters don't affect the subsequent analysis but the dictionary structure needs to be there. Let me know how you think we could change it! :))

niklasmueboe replied:
So do they always assume delaunay triangulation as input for the neighbor structure? Because they also do this n-th percentile trimming of the neighbor graph which does not necessarily make sense for some other methods, e.g. anything that is based of neighbors on a regular grid which is usually used for Visium for example.

Jieran-S replied:
Ohh that's a good point! Sorry I missed that earlier. In their documentation, they only showcase the example of CosMx and they use `delaunay=True`. But in their [paper](https://www.nature.com/articles/s41588-023-01588-4#Sec7) they did mention that removing long links is unnecessary for regular structure technology. 
I will make it optional based on the `technology` option. If `technology==Visium`, I will skip the distance calculation and the `remove_long_links` part, as the distance matrix is only needed for`remove_long_links`.

niklasmueboe replied:
It should also be done for "ST" not only "Visium".

The issue I see that it is technically not only dependent on the technology but also on the neighbor definition. e.g. you could define neighbors based on distance and then doing a trimming on the 99th percentile does not make much sense.

Not quite sure what the best approach is for now.

Jieran-S replied:
Since the tirmming only makes sense when delaunay triangulation is applied to construct the neighborhood graph (hence not depending on the distance), I can try to implement some codes to check if the neighbor graphs provided fulfills delaunay criteria and proceed on trimming based on that. Do you think that'd be a way to work around it? :))

Jieran-S committed:
remove private import

Jieran-S commented:
Another thing is, cellcharter uses scVI for dimensional reduction and batch correction, it was mentioned in their paper to be an important part. I did not implement that part as we will already provide a `dim_red` slot. Should I maybe make it an option for the scVI batch correction and use config file to activate it, or should I assume our preprocessing will test out all possibilities?

Jieran-S committed:
optional link removal based on techs

Jieran-S committed:
update docs

niklasmueboe commented:
> Another thing is, cellcharter uses scVI for dimensional reduction and batch correction, it was mentioned in their paper to be an important part. I did not implement that part as we will already provide a `dim_red` slot. Should I maybe make it an option for the scVI batch correction and use config file to activate it, or should I assume our preprocessing will test out all possibilities?

bacth correction should nly be necessary for multi-sample analysis which we haven't implemented for now. And regarding dim-red I think we can always add scVI as additional preprocessing step alternatively to PCA, ...

So I think we can leave it as is for now

Jieran-S committed:
update raw data input

Jieran-S commented:
Hi, I just talked to the author of cellcharter and updated our implementation: 
1. Added cellcharter support for raw data input. This is important as the dimensional reduction embedding provided by scVI is essential to cellcharter performance. 
2. Simply restrict the "remove_long_link" function to technology types(i.e. not applied to Visium and ST). The author mentioned that removing the top 1% or not actually do not affect the performance much. Even if the graph was not generated via delaunay triangulation, removing the top 1% wouldnt have too much of an impact.

Jieran-S committed:
update yml format

Jieran-S committed:
Add optargs for cellcharter

Jieran-S committed:
cc raw count options

Jieran-S commented:
Change CC to accepting raw count input only. So:

1. Delaunay neighbor graph cut issue resolved
2. scVI is important to cellcharter. Can't be replaced by other dim-red preprocessing.

niklasmueboe commented on method/CellCharter/CellCharter.py:
```diff
@@ -0,0 +1,198 @@
+)
+
+model = scvi.model.SCVI(adata)
+model.train(early_stopping=True, enable_progress_bar=True)
```
Is this an interactive progress bar? Then it should probably be set to false?

Jieran-S replied:
Tune them off! :))

niklasmueboe commented on method/CellCharter/CellCharter.py:
```diff
@@ -0,0 +1,198 @@
+    convergence_tolerance=config["convergence_tolerance"], 
+    covariance_regularization=config["covariance_regularization"]
+    # if running on gpu
+    #trainer_params=dict(accelerator='gpu', devices=1)
```
So can we currently use GPU for this tool?

Jieran-S replied:
Yes -- I added a cuda of GPU availability now.

Jieran-S committed:
progress bar and gpu options

Jieran-S committed:
oops forgot a comma
